cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -21,40 +22,113 @@ import Data.Ord
+sendMsgQuery :: Monad m => query a -> LocalStateQueryExpr block point query r m a
+sendMsgQuery q = LocalStateQueryExpr . ContT $ \f -> pure $
   Net.Query.SendMsgQuery q $
     Net.Query.ClientStQuerying
     { Net.Query.recvMsgResult = f
     }
 
+-- | Execute a local state query expression.
+executeQueryLocalState
+  :: LocalNodeConnectInfo CardanoMode
+  -> Maybe ChainPoint
+  -> (NodeToClientVersion -> LocalStateQueryExpr (BlockInMode CardanoMode) ChainPoint (QueryInMode CardanoMode) () IO a)
+  -> IO (Either AcquireFailure (Maybe a))
+executeQueryLocalState connectInfo mpoint f = do
+  resultVarQueryTipLocalState <- newEmptyTMVarIO
+  waitResult <- pure $ sequence <$> readTMVar resultVarQueryTipLocalState
+
+  connectToLocalNodeWithVersion
+    connectInfo
+    (\ntcVersion ->
+      LocalNodeClientProtocols
+      { localChainSyncClient    = NoLocalChainSyncClient
+      , localStateQueryClient   = Just $ setupLocalStateQueryExpr waitResult mpoint ntcVersion resultVarQueryTipLocalState (f ntcVersion)
+      , localTxSubmissionClient = Nothing
+      }
+    )
+
+  atomically waitResult
+
+-- | Execute a local state query expression concurrently with a chain sync.
+executeQueryLocalStateWithChainSync
+  :: LocalNodeConnectInfo CardanoMode
+  -> Maybe ChainPoint
+  -> (NodeToClientVersion -> LocalStateQueryExpr (BlockInMode CardanoMode) ChainPoint (QueryInMode CardanoMode) () IO a)
+  -> IO (ChainTip, Either AcquireFailure (Maybe a))
+executeQueryLocalStateWithChainSync connectInfo mpoint f = do
+  resultVarQueryTipLocalState <- newEmptyTMVarIO
+  resultVarChainTip <- newEmptyTMVarIO
+
+  waitResult <- pure $ (,)
+    <$> readTMVar resultVarChainTip
+    <*> (sequence <$> readTMVar resultVarQueryTipLocalState)
+
+  connectToLocalNodeWithVersion
+    connectInfo
+    (\ntcVersion ->
+      LocalNodeClientProtocols
+      { localChainSyncClient    = LocalChainSyncClient $ chainSyncGetCurrentTip waitResult resultVarChainTip
+      , localStateQueryClient   = Just $ setupLocalStateQueryExpr waitResult mpoint ntcVersion resultVarQueryTipLocalState (f ntcVersion)
+      , localTxSubmissionClient = Nothing
+      }
+    )
+
+  atomically waitResult
+
+  where
+    chainSyncGetCurrentTip
+      :: STM b
+      -> TMVar  ChainTip
+      -> ChainSyncClient (BlockInMode mode) ChainPoint ChainTip IO ()
+    chainSyncGetCurrentTip waitDone tipVar =
+      ChainSyncClient $ pure clientStIdle
+      where
+        clientStIdle :: Net.Sync.ClientStIdle (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStIdle =
+          Net.Sync.SendMsgRequestNext clientStNext (pure clientStNext)
+
+        clientStNext :: Net.Sync.ClientStNext (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStNext = Net.Sync.ClientStNext
+          { Net.Sync.recvMsgRollForward = \_block tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
+              void $ atomically waitDone
```

Comment:
What is `void $ atomically waitDone` doing here? Is it necessary?

Reply:
It's waiting for all the queries in all the protocols to be "done".  If we exit when in one protocol before another protocol is done we get an STM deadlock because the other protocol is aborted and never finishes.

Reply:
Done!